File: web/web-lobby/src/main/java/com/yazino/web/api/TournamentController.java
package com.yazino.web.api;

import com.yazino.platform.tournament.TournamentView;
import com.yazino.web.data.TournamentViewRepository;
import com.yazino.web.domain.TournamentDetailView;
import com.yazino.web.domain.TournamentLobbyCache;
import com.yazino.web.session.LobbySession;
import com.yazino.web.session.LobbySessionCache;
import com.yazino.web.util.PlayerFriendsCache;
import com.yazino.web.util.WebApiResponses;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.math.BigDecimal;

import static javax.servlet.http.HttpServletResponse.SC_INTERNAL_SERVER_ERROR;
import static javax.servlet.http.HttpServletResponse.SC_UNAUTHORIZED;
import static org.apache.commons.lang3.Validate.notBlank;
import static org.apache.commons.lang3.Validate.notNull;

@Controller("apiTournamentController")
@RequestMapping("/api/1.0/tournament/*")
public class TournamentController {
    private static final Logger LOG = LoggerFactory.getLogger(TournamentController.class);

    private final LobbySessionCache lobbySessionCache;
    private final TournamentLobbyCache tournamentLobbyCache;
    private final TournamentViewRepository tournamentViewRepository;
    private final PlayerFriendsCache playerFriendsCache;
    private final WebApiResponses webApiResponses;

    @Autowired
    public TournamentController(final LobbySessionCache lobbySessionCache,
                                final TournamentLobbyCache tournamentLobbyCache,
                                final TournamentViewRepository tournamentViewRepository,
                                final PlayerFriendsCache playerFriendsCache,
                                final WebApiResponses webApiResponses) {
        notNull(lobbySessionCache, "lobbySessionCache may not be null");
        notNull(tournamentLobbyCache, "tournamentLobbyCache may not be null");
        notNull(tournamentViewRepository, "tournamentViewRepository may not be null");
        notNull(playerFriendsCache, "playerFriendsCache may not be null");
        notNull(webApiResponses, "webApiResponses may not be null");

        this.lobbySessionCache = lobbySessionCache;
        this.tournamentLobbyCache = tournamentLobbyCache;
        this.tournamentViewRepository = tournamentViewRepository;
        this.playerFriendsCache = playerFriendsCache;
        this.webApiResponses = webApiResponses;
    }

    @RequestMapping(value = "/next/by-variation/{gameType}", method = RequestMethod.GET)
    public void nextTournamentsForAllVariations(@PathVariable("gameType") final String gameType,
                                                final HttpServletRequest request,
                                                final HttpServletResponse response) throws IOException {
        notBlank(gameType, "gameType may not be null/blank");
        notNull(request, "request may not be null");
        notNull(response, "response may not be null");

        BigDecimal playerId = null;
        try {
            final LobbySession currentSession = lobbySessionCache.getActiveSession(request);
            if (currentSession == null) {
                webApiResponses.writeError(response, SC_UNAUTHORIZED, "No session");
                return;
            }
            playerId = currentSession.getPlayerId();

            webApiResponses.writeOk(response, tournamentLobbyCache.getNextTournamentsForEachVariation(playerId, gameType));

        } catch (Exception e) {
            LOG.error("Failed to get next tournaments for game type {} and player {}", gameType, playerId, e);
            webApiResponses.writeError(response, SC_INTERNAL_SERVER_ERROR,
                    String.format("Failed to get next tournaments for game type %s and player %s", gameType, playerId));
        }
    }

    @RequestMapping(value = "/{tournamentId}", method = RequestMethod.GET)
    public void tournamentDetailFor(@PathVariable("tournamentId") final BigDecimal tournamentId,
                                    final HttpServletRequest request,
                                    final HttpServletResponse response) throws IOException {
        notNull(tournamentId, "tournamentId may not be null");
        notNull(request, "request may not be null");
        notNull(response, "response may not be null");

        BigDecimal playerId = null;
        try {
            final LobbySession currentSession = lobbySessionCache.getActiveSession(request);
            if (currentSession == null) {
                webApiResponses.writeError(response, SC_UNAUTHORIZED, "No session");
                return;
            }
            playerId = currentSession.getPlayerId();

            final TournamentView tournamentView = tournamentViewRepository.getTournamentView(tournamentId);
            if (tournamentView != null) {
                webApiResponses.writeOk(response, new TournamentDetailView(tournamentView, playerId, playerFriendsCache.getFriendIds(playerId)));

            } else {
                webApiResponses.writeNoContent(response, HttpServletResponse.SC_NOT_FOUND);
            }

        } catch (Exception e) {
            LOG.error("Failed to get tournament detail for {} with player {}", tournamentId, playerId, e);
            webApiResponses.writeError(response, SC_INTERNAL_SERVER_ERROR,
                    String.format("Failed to get tournament detail for %s with player %s", tournamentId, playerId));
        }
    }

}

File: platform/control-centre/src/main/scala/strata/server/lobby/controlcentre/controller/TournamentController.scala
package strata.server.lobby.controlcentre.controller

import org.springframework.stereotype.Controller
import org.springframework.beans.factory.annotation.Autowired
import strata.server.lobby.controlcentre.repository.{TournamentRepository, TournamentVariationRepository}
import org.apache.commons.lang3.Validate._
import org.springframework.web.bind.annotation._
import org.springframework.web.bind.WebDataBinder
import strata.server.lobby.controlcentre.util.BigDecimalPropertyEditorSupport
import org.springframework.beans.propertyeditors.CustomDateEditor
import java.text.SimpleDateFormat
import java.util
import strata.server.lobby.controlcentre.validation.TournamentValidator
import org.springframework.web.servlet.ModelAndView
import javax.servlet.http.HttpServletResponse
import strata.server.lobby.controlcentre.form.TournamentForm
import scala.Array
import org.springframework.validation.BindingResult
import com.yazino.platform.tournament.TournamentService
import java.math

@Controller
class TournamentController @Autowired()(val tournamentRepository: TournamentRepository,
                                        val tournamentVariationRepository: TournamentVariationRepository,
                                        val tournamentService: TournamentService,
                                        val tournamentValidator: TournamentValidator) {
    notNull(tournamentRepository, "tournamentRepository may not be null")
    notNull(tournamentVariationRepository, "tournamentVariationRepository may not be null")
    notNull(tournamentService, "tournamentService may not be null")
    notNull(tournamentValidator, "tournamentValidator may not be null")

    @InitBinder def initBinder(binder: WebDataBinder) {
        binder.registerCustomEditor(classOf[BigDecimal], new BigDecimalPropertyEditorSupport())

        val editor = new CustomDateEditor(new SimpleDateFormat("yyyy-MM-dd'T'HH:mm"), true)
        binder.registerCustomEditor(classOf[util.Date], editor)
    }

    @ModelAttribute("variations")
    def variations: util.TreeMap[math.BigDecimal, String] = {
        val variationsToNames = new util.TreeMap[math.BigDecimal, String]()
        tournamentVariationRepository.list().foreach {
            item => variationsToNames.put(item._1.underlying(), item._2)
        }
        variationsToNames
    }

    @RequestMapping(Array("tournament/tournament/list"))
    def list: ModelAndView = listAtPage(1)

    @RequestMapping(Array("tournament/tournament/list/{page}"))
    def listAtPage(@PathVariable("page") page: Int = 1): ModelAndView = {
        new ModelAndView("tournament/tournament/list")
            .addObject("tournaments", tournamentRepository.findAll(page - 1, 20))
            .addObject("cannotCancelStatuses", TournamentForm.cannotCancelStatuses)
    }

    @RequestMapping(Array("tournament/tournament/show/{tournamentId}"))
    def show(@PathVariable("tournamentId") tournamentId: Int,
             response: HttpServletResponse): ModelAndView = {
        val tournament = tournamentRepository.findById(BigDecimal(tournamentId))

        if (tournament.isDefined) {
            new ModelAndView("tournament/tournament/show")
                .addObject("tournament", new TournamentForm(tournament.get))
        } else {
            response.sendError(HttpServletResponse.SC_NOT_FOUND)
            null
        }
    }

    @RequestMapping(Array("tournament/tournament/create"))
    def create(): ModelAndView =
        new ModelAndView("tournament/tournament/create")
            .addObject("tournament", new TournamentForm())

    @RequestMapping(value = Array("tournament/tournament/save"), method = Array(RequestMethod.POST))
    def save(@ModelAttribute("tournament") form: TournamentForm,
             bindingResult: BindingResult): ModelAndView = {
        notNull(form, "form may not be null")

        tournamentValidator.validate(form, bindingResult)
        if (bindingResult.hasErrors) {
            new ModelAndView("tournament/tournament/create")
                .addObject("tournament", form)

        } else {
            val tournamentId = tournamentService.createTournament(
                form.toTournament.toPlatform(tournamentVariationRepository))

            new ModelAndView("redirect:/tournament/tournament/show/%s".format(tournamentId))
        }
    }

    @RequestMapping(Array("tournament/tournament/cancel/{tournamentId}"))
    def cancel(@PathVariable("tournamentId") tournamentId: Int): ModelAndView =
        if (tournamentService.cancelTournament(BigDecimal(tournamentId).underlying())) {
            list.addObject("message", "Tournament %s has been cancelled.".format(tournamentId))
        } else {
            list.addObject("message", "Tournament %s has not been cancelled.".format(tournamentId))
        }
}
